### PR TITLE
fix: support propagating explicit cast of array

### LINKED ISSUE #6077: bug(binder): explicit cast of array constructor should propagate context inside
### Describe the bug

This is an edge case only when mixing elements of different type categories via an outer explicit cast. (low priority)

Working as expected:
```
select array[1, 2.5];  -- same type categories where implicit cast applies
select array[1, true];  -- expected error due to different type categories
select array[1::bool, true];  -- okay when explicit cast applied inside
```

The following is allowed by PostgreSQL but rejected in RisingWave, because `array[1, true]` is not a valid subexpr.
```
select array[1, true]::bool[];
```

### To Reproduce

```
select array[1, true]::bool[];
```

### Expected behavior

```
test=# select array[1, true]::bool[];
 array 
-------
 {t,t}
(1 row)
```

### Additional context

There was a similar issue for `insert values`, where extra external type info is preferred over usage of `align_types`:

https://github.com/risingwavelabs/risingwave/blob/87bf20c79e91d58dae3e7c19be055aec718455f2/src/frontend/src/binder/insert.rs#L56-L74

Another similar and related issue we already solved is the literal empty array, where `array[]` is an invalid subexpr but `array[]::int[]` as a whole is okay. We can just refactor/extend this part to solve this issue.

https://github.com/risingwavelabs/risingwave/blob/87bf20c79e91d58dae3e7c19be055aec718455f2/src/frontend/src/binder/expr/mod.rs#L391-L392

### PR DESCRIPTION
**Pull request by ZENOTME**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

as title, fix #6077 

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
#6077

### CODE REVIEW

File: src/frontend/src/binder/expr/value.rs [hunk 1]
```diff
@@ -110,6 +112,39 @@ impl Binder {
...
+        let mut exprs = exprs
+            .into_iter()
+            .map(|e| self.bind_expr(e))
+            .collect::<Result<Vec<ExprImpl>>>()?;
+        // Cast to the same type.
+        for e in &mut exprs {
+            let mut dummy = ExprImpl::literal_bool(false);
+            swap(&mut dummy, e);
+            *e = dummy.cast_explicit(ty.clone())?;
+        }
```
Comment (xiangjinwu) [inline src/frontend/src/binder/expr/value.rs:136]: The propagation needs to happen recursively for multi-dimensional arrays. For example `select array[array[1, true]]::bool[][];`

File: src/frontend/src/binder/expr/value.rs [hunk 2]
```diff
@@ -110,6 +112,39 @@ impl Binder {
         Ok(expr)
     }
 
+    pub(super) fn bind_array_cast(&mut self, exprs: Vec<Expr>, ty: DataType) -> Result<ExprImpl> {
+        if exprs.is_empty() {
+            return Ok(FunctionCall::new_unchecked(
+                ExprType::Array,
+                vec![],
+                // Treat `array[]` as `varchar[]` temporarily before applying cast.
```
Comment (xiangjinwu) [inline src/frontend/src/binder/expr/value.rs:120]: This comment is no longer valid. We already use the actual `ty` here.

File: src/frontend/src/binder/expr/mod.rs [hunk 3]
```diff
@@ -385,21 +385,9 @@ impl Binder {
...
-                },
-            )
-            .into()
-        } else {
+        let lhs = if let Expr::Array(exprs) = &expr && let AstDataType::Array(data_type) = &data_type {
+            self.bind_array_cast(exprs.to_owned(),bind_data_type(data_type)?)?
+        }else {
             self.bind_expr(expr)?
         };
         lhs.cast_explicit(bind_data_type(&data_type)?)
```
Comment (xiangjinwu) [inline src/frontend/src/binder/expr/mod.rs:393]: This is no longer needed for the `if` case and can be inside `else`

**Comment by fuyufjh** [comment]
Any updates?

**Review by fuyufjh — APPROVED** [review]

**Comment by ZENOTME** [comment]
> Any updates?

To add the related e2e test, I will finish it after #6124.

**Review by xiangjinwu — APPROVED** [review]

**Pull request merged**